### ISSUE TITLE
Rojo is too hard to install, and should ship OS-specific installers

### ISSUE THREAD

**Issue by kennethloeffler** [rojo-rbx/rojo#676]
Rojo is too hard to install. This hurts Rojo's adoption because most new users simply quit and never come back before they even get it installed. Software that requires the user to manually place an executable and edit the PATH variable is rare... software that wants to succeed, anyway. As someone who actively tries to convince others to use Rojo, it's kind of embarrassing. A new user shouldn't have to know what the PATH variable is or know how to edit a special text file just to get the damn thing on their system.

The typical user is very familiar with opening a `.msi` or `.pkg` file to install new software. These methods of installation are ubiquitous, anyone who seriously uses Windows or macOS knows how to do it, and anything different is going to seem weird, complicated, and difficult to most users.

I know that there have been discussions and arguments about the packaging and distribution of standalone Rojo before (i.e. not installed via the VS Code extension [and/or using a bespoke installation method that seems to confuse new people pretty often.. but I digress]) . I've seen a lot of consternation about packaging for different Linux distributions. The way I see it, worrying about them makes no sense when the app that we're trying to integrate with (Roblox Studio) *doesn't even run on those platforms*. If someone really wants Rojo on their favorite backwoods package manager, they can do it themselves. These "advanced users" should have the skills to do so, right? It's their problem, not ours.

I also know that we have historically been cautious about doing anything platform specific on this project, but I think we should bite this bullet. Most of our users are on one of two platforms, and it's really not too much work to create Windows and macOS installers. I think we're way past due on this. I have experience with the WiX Toolset (WiX v3 just released and it's easier to pick up and use than ever), and I'm willing to learn how to do macOS installers if no one else is. 

Is this something we want?

**Comment by LastTalon** [comment]
I don't have an issue with having an installer per se, but I don't really see the need for a complicated installer either. It's a single binary and one line of configuration. The configuration never changes. The binary can be updated, but the same installer can always pull the latest version.

Aftman and foreman already do this. Aftman can even self-install. So to me the correct route would seem to be just to supply a short installation script that either downloads the binary and configures the path itself, or uses aftman to do this. MSI is for merging larger applications that have much more complicated updates and I don't really see that as necessary for moving and replacing a single binary file. Same for other packaging formats I've encountered.

This installation method is really common for command line tools. See rustup installation. All of that said, if you want to take on making those particular msi/pkg installers I don't really see a problem with it either. Perhaps we can kick off discussion on why you think those platform-specific installers are a good choice. For my part, I think tools like aftman or an install script are more useful for command line tools like this because they're a good starting point that works for all platforms. These tools typically get installed somewhere like `~/bin` by users anyway rather than being system managed applications.

**Comment by kennethloeffler** [comment]
> Perhaps we can kick off discussion on why you think those platform-specific installers are a good choice.
I think that a new user might have never used a command line tool (much less know how to set it up properly), but they will almost always know how to double-click an installer they've downloaded from the internet. Guess what the very first thing this kind of user will do with the Rojo binary once they get their hands on it.

> Aftman and foreman already do this [...] Aftman can even self-install. This installation method is really common for command line tools. See rustup installation.
I understand and appreciate the value of tools like Foreman and Aftman. I use them every day, and will continue to do so. I understand there's all kinds of installation methods developers in the wider software world use that work perfectly well for them. Still, I think we must do a better job of playing into *our* users' expectations. They do not expect a multi-step process that begins "first, install the installer..." I think "predictable" is more important than "pure," especially when we're talking installation.

> I don't have an issue with having an installer per se, but I don't really see the need for a complicated installer either. It's a single binary and one line of configuration [...] MSI is for merging larger applications that have much more complicated updates and I don't really see that as necessary for moving and replacing a single binary file. Same for other packaging formats I've encountered.
Standard installation solutions contain a lot of complexity because they need to support arbitrary installations on arbitrarily configured systems, and most importantly, *in a way that users expect* (perhaps registering file associations, creating an entry in add or remove programs, not leaving the system dirty after uninstallation, providing a way to uninstall *at all*, etc.). This is an entire field of software engineering. It's actually pretty dang hard to do it even close to correctly, and I don't see a good reason for why we should shoulder the burden. We are already spread thin.

I disagree with the assertion that the standard methods are only meant particular kinds of applications, or that the installer must necessarily be complicated just because the underlying technologies are complicated. It's also entirely possible we'll want more out of installations in the future: for example, once Rojo dynamically loads reflection databases, what if we want to avoid downloading every time, and instead store them locally? Where would be a good place to keep them? This will soon cease to be hypothetical.

---

With all this in mind, I think that if we go to improve the standalone installation experience, AND we stick with the bespoke solutions, then we'll end up reimplementing what the standard solutions have already done for decades. That looks like a bad deal to me.

Windows Installer/MSI are definitely not the best ever pieces of tech, but that ship sailed long ago. They do their jobs as they're supposed to, and I don't believe our goal is to start a deployment revolution (at least, it really shouldn't be). Rojo exists in a context where most users are on one of two platforms. They come with certain expectations. Categorically refusing to accommodate them will cause many of them to abandon the software. It seems pretty cut-and-dry.

**Comment by CompeyDev** [comment]
I've done [similar work for lune](https://github.com/CompeyDev/lune-packaging), and although I don't think I'll be able to dedicate vast extents of time to maintain multiple platform packaging for Rojo, I think I'll be able to handle packaging for platforms like the AUR or Homebrew.

Although both of these platforms aren't GUI-like which are the simplest are most used, they are still quite a bit more easier to use, since they usually only involve executing a single command.

Lmk if you guys are interested in this.

**Comment by LastTalon** [comment]
To be clear, because I guess I never really followed up on this:

> although I don't think I'll be able to dedicate vast extents of time to maintain multiple platform packaging for Rojo, I think I'll be able to handle packaging for platforms like the AUR or Homebrew.

This seems good to me. My main concern is not whether or what standard installation solutions are, but the complexity of maintaining them. Complexity is the reason that even relatively simple packaging is often done by individual package maintainers and not the projects themselves. It's just too much additional work and complexity.

That's why I said:

> All of that said, if you want to take on making those particular msi/pkg installers I don't really see a problem with it either.

If you're willing to take on that complexity I don't have a problem with it. I just don't think it's a one-and-done thing. It's a commitment to ongoing maintenance of that installer. Once it's an official installation method you've basically locked in that expectation for users.

**Comment by CompeyDev** [comment]
To be fair, most packaging is one time, except for things like dpkg/aur (usually pushing to repositories). For instance, you'd probably never need to update an MSI's configs except for a 'version' variable, which can be often updated in CI. I feel like stuff like native installers should be provided by Rojo, officially; but we can have repository specific packages community maintained.

BTW, brew casks require **ZERO** maintenance once you push them to Homebrew core, there's hundreds or thousands of people which actively maintain and update them, without any input from the initial publisher.

**Comment by Dekkonot** [comment]
So, I've thought about this off and on. I definitely see the appeal of installers because it would remove some of the friction from users, but I don't think it's where we should focus our energy.

Here's my reasoning: what happens after someone installs Rojo? If someone is using VSCode, there's really no reason we couldn't download Rojo for them in the same way tools like Selene and StyLua do. It's easier and removes all external downloads for those users.

If they're using the CLI, I think they're going to have to get comfortable putting stuff on their path one way or another unless Rojo is the only tool they use. Aftman/Foreman is one option for this, but I honestly think simplicity is the best option here: we should provide step-by-step instructions for adding things to paths along with basic troubleshooting steps.

Our core audience for installers would be Roblox users who have never touched a CLI before. It does them a disservice to try to hide the process of adding one to a PATH from them when we could instead just educate them.

**Comment by CompeyDev** [comment]
True, however the other use-case is to simplify repeated tasks. It's much more simpler for me to install Rojo in a new installation of Arch Linux with an AUR command, instead of having to download it from GitHub and put it in my path or installing Aftman, then installing Rojo.

But yes, I feel like it's a bit unnecessary to have GUI installers for a CLI program. At max, a shell/batch script should suffice for ease of use.

**Comment by CompeyDev** [comment]
I've not published [rojo](https://aur.archlinux.org/packages/rojo), [rojo-git](https://aur.archlinux.org/packages/rojo-git), and [rojo-bin](https://aur.archlinux.org/packages/rojo-bin) to the AUR. Is there anywhere on the docs that I might need to add these to?

**Comment by CompeyDev** [comment]
Also, as for `pkg` and `msi` installers, I found a crate which makes it fairly simple to generate such OS-specific installers. See [burtonageo/cargo-bundle](https://github.com/burtonageo/cargo-bundle).
